src/compiler/checker.ts

```diff
@@ -4450,6 +4518,10 @@ namespace ts {
             return type.resolvedApparentType;
         }
 
+        function getApparentTypeOfSpread(type: SpreadType) {
+            return getApparentType(type.right);

```

Comment:
This implementation is simplified from reality, but probably not often enough to matter. First, remember that spread types are only created if there's an type parameter embedded in it [1]. So we're talking about types like `T ... U, { x } ... T or T ... { x }`

This code is correct for `T ... U and { x } ... T`. For example [2], if `*T = { x: string }` and `*U = { y: number }` then you can only say for sure that `*(T ... U) = { y: number }` because when `U` is instantiated, it might have `{ x: number }`, meaning that the instantiated type would actually be `{ x: number, y: number }`. And if `U` ends up with `{ x?: number }`, then the final type would have `x: number | string`.

Same reasoning applies to `{ x } ... T` &mdash; `T` might contain `x` with a different type.

That leaves `T ... { x }` which could actually distribute the apparent type down, but I don't think this case is worth enough in practise to justify the additional code. You can go back a few commits to see what the code looked like, although it would need to be surrounded with `if (type.right & TypeFlags.Object) ... else return getApparentType(type.right)`.

[1] Or an intersection with a type parameter embedded in *it*.
[2] I'll use the syntax `*T` to shorten `getApparentType(T)`.